### ISSUE TITLE
Unable to find some shared libraries of executable

### ISSUE THREAD

**Issue by rodolfo-picoreti** [conan-io/conan#3721]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

conan: v1.8.1
os: ubuntu16.04 / ubuntu18.04
I've read: https://docs.conan.io/en/latest/howtos/manage_shared_libraries.html

Hi, I am trying to build an executable that has some shared libraries as dependencies. On ubuntu 16.04 (gcc 5) everything works as expected. I can build and run them with no problems. However on ubuntu 18.04 (gcc 7) the artifact is built with no problem, although when trying to run the executable the linker cant find some of the libraries. When inspecting the output of ldd I see that part of the packages are found but the other part is not.

```
libopencv_core.so.3.4 => /home/hodor/.conan/data/opencv/3.4.2/is/stable/package/.../libopencv_core.so.3.4 (0x00007f9b70922000)
libtbbmalloc_proxy.so.2 => /home/hodor/.conan/data/TBB/4.4.4/conan/stable/package/.../libtbbmalloc_proxy.so.2 (0x00007f9b7071f000)
libboost_filesystem.so.1.66.0 => /home/hodor/.conan/data/boost/1.66.0/conan/stable/package/.../libboost_filesystem.so.1.66.0 (0x00007f9b70065000)
... [The libraries below are also built with conan and located in /home/hodor/.conan/data/...]
libprometheus-cpp.so => not found      
libSimpleAmqpClient.so.2 => not found
libzipkin.so.0 => not found
```

Any idea on what could be happening? Why only part of the libraries are found on ubuntu1804 (gcc7)?

**Comment by memsharded** [comment]
Yes, this might happen because the opencv recipe you are using is implementing a ``system_requirements()`` method to install some system packages that are required for opencv. It is possible that the recipes were developed and tested in older distros.

However, it is a bit weird, the only system dep installed by the 3.X opencv recipe is Gtk: https://github.com/conan-community/conan-opencv/blob/d5f98b3f7e613fcdca1b0b0332edf8a0886a4291/conanfile.py#L91.

So I have no idea why those prometheus, SimpleAmqpClient, zipkin dependencies could be coming from. If you could please share some repo or recipe to reproduce this error, that would help. Thanks!

**Comment by rodolfo-picoreti** [comment]
Hey @memsharded 

Sorry for not explaining the issue properly. Opencv, prometheus, SimpleAmqpClient and zipkin are libraries that come from custom recipes at https://github.com/labviros/is-packages.
I am planning on creating a minimal reproducible example, but I didnt have the time yet.
For now, if you want, you can use https://github.com/labviros/is-frame-transformation to reproduce the error. Due to #3729 I am not able to build docker images based on ubuntu16/18 to see if I can reproduce it there as well. I'll try it later today.

off-topic: I was not aware of the 3.4 version in the official opencv package, I'll try to contribute some changes there later.

**Comment by rodolfo-picoreti** [comment]
I've tested https://github.com/labviros/is-frame-transformation using a docker container based on ubuntu16.04 and 18.04 images and I was able to reproduce the problem there, that is, in 18.04 (gcc7) when executing the produced binary only part of the shared libraries that come from conan are not found.

I was able to reproduce the problem on a smaller scale:
https://github.com/rodolfo-picoreti/conan-issue-3721

**Comment by rodolfo-picoreti** [comment]
@memsharded, in summary I was able to [reproduce](https://github.com/rodolfo-picoreti/conan-issue-3721) the problem by having the executable depend on a single [library](https://github.com/labviros/is-msgs/blob/master/conanfile.py) that itself depends on three others (Boost, fmt and protobuf). 

After building the executable, if I run **ldd** on it:
On Ubuntu 16:04 (gcc5):
```
linux-vdso.so.1 =>  (0x00007ffdc1ff6000)
libis-msgs.so => /root/.conan/data/is-msgs/1.1.8/is/stable/package/c2ece6c6387a7d67d3d411c0d35deb51f15ce679/lib/libis-msgs.so (0x00007f3f934af000)
libprotobuf-lite.so.3.6.1 => /root/.conan/data/protobuf/3.6.1/bincrafters/stable/package/ed432596ad457ff90665e1ac0e3d559db99e5728/lib/libprotobuf-lite.so.3.6.1 (0x00007f3f93258000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f3f92ed6000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f3f92cc0000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f3f928f6000)
libprotobuf.so.3.6.1 => /root/.conan/data/protobuf/3.6.1/bincrafters/stable/package/ed432596ad457ff90665e1ac0e3d559db99e5728/lib/libprotobuf.so.3.6.1 (0x00007f3f9244c000)
libboost_filesystem.so.1.66.0 => /root/.conan/data/boost/1.66.0/conan/stable/package/1dda960c639ec6bd5760b2fa6bb180d91cb383a8/lib/libboost_filesystem.so.1.66.0 (0x00007f3f92231000)
libfmt.so.5 => /root/.conan/data/fmt/5.1.0/bincrafters/stable/package/79fff9d59ddc62bb06d68e6f3dfaf1166fde4418/lib/libfmt.so.5 (0x00007f3f91fec000)
libboost_system.so.1.66.0 => /root/.conan/data/boost/1.66.0/conan/stable/package/1dda960c639ec6bd5760b2fa6bb180d91cb383a8/lib/libboost_system.so.1.66.0 (0x00007f3f91de8000)
/lib64/ld-linux-x86-64.so.2 (0x00007f3f937b2000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f3f91bcb000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f3f918c2000)
```

On Ubuntu 18:04 (gcc7):
```
linux-vdso.so.1 (0x00007ffc2abad000)
libis-msgs.so => /root/.conan/data/is-msgs/1.1.8/is/stable/package/a89aa91d8eeddc75020a5c000f32f0084d173216/lib/libis-msgs.so (0x00007f28426b1000)
libprotobuf-lite.so.3.6.1 => /root/.conan/data/protobuf/3.6.1/bincrafters/stable/package/ce3b1f5942a6405e42e407c01438f95ae599df8f/lib/libprotobuf-lite.so.3.6.1 (0x00007f284245a000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f28420d1000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f2841eb9000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2841ac8000)
/lib64/ld-linux-x86-64.so.2 (0x00007f2842bb2000)
libprotobuf.so.3.6.1 => not found
libboost_filesystem.so.1.66.0 => not found
libfmt.so.5 => not found
libboost_system.so.1.66.0 => not found
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f28418a9000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f284150b000)

```

**Comment by rodolfo-picoreti** [comment]
Hi @memsharded, any update on this? Is the explanation above good enough to reproduce the problem?

**Comment by nickaleks** [comment]
Any update on this? I've been running into the same issue for a while.

**Comment by resqsx** [comment]
I have the same issue and opened a test project for this.
See also
https://stackoverflow.com/questions/61210938/conan-boost-shared-library-not-found
and
https://github.com/resqsx/conan-boost.

It would be nice to get a solution for this.

**Comment by resqsx** [comment]
It looks like the transitive dependencies of the boost libraries can not be found by the loader.
This is because since Ubuntu 18.04. cmake will install the libraries with **RUNPATH**.
But the application loader doesn't use **RUNPATH** to resolve transitive dependencies. 
But **RPATH** does the job.
One way would be to build the boost libs to find there dependencies with RUNPATH.
Which I would recommend to do.

To solve the problem for now you can tell cmake to use RPATH like so.
`set(CMAKE_EXE_LINKER_FLAGS "-Wl,--disable-new-dtags")`

**Comment by nickaleks** [comment]
@resqsx The solution you provided seems to be working. I am wondering how building with RUNPATH should be achieved - should all the deps just be rebuilt using --enable-new-dtags, such as on Ubuntu 18.04?

Also should Conan do anything about it? Maybe emit a warning if the package has been built without RUNPATH on a system where it is expected?

**Comment by AbrilRBS** [comment]
Closing as part of our cleanup for Conan 1-only issues, which is no longer supported
